File: core/src/main/java/vn/loitp/views/textview/typewritertextview/lib/TypeWriterTextView.java
package vn.loitp.views.textview.typewritertextview.lib;

import android.content.Context;
import android.os.Handler;
import android.util.AttributeSet;
import android.widget.TextView;

/**
 * Created by www.muathu@gmail.com on 11/2/2017.
 */

public class TypeWriterTextView extends TextView {

    private CharSequence mText;
    private int mIndex;
    private long mDelay = 150; // in ms

    public TypeWriterTextView(Context context) {
        super(context);
    }

    public TypeWriterTextView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    private Handler mHandler = new Handler();

    private Runnable characterAdder = new Runnable() {

        @Override
        public void run() {
            setText(mText.subSequence(0, mIndex++));

            if (mIndex <= mText.length()) {
                mHandler.postDelayed(characterAdder, mDelay);
            }
        }
    };

    public void animateText(CharSequence txt) {
        mText = txt;
        mIndex = 0;

        setText("");
        mHandler.removeCallbacks(characterAdder);
        mHandler.postDelayed(characterAdder, mDelay);
    }

    public void setCharacterDelay(long m) {
        mDelay = m;
    }
}

File: core/src/main/java/com/views/textview/typewritertextview/TypeWriterTextView.kt
package com.views.textview.typewritertextview

import android.content.Context
import android.os.Handler
import android.util.AttributeSet

import androidx.appcompat.widget.AppCompatTextView

/**
 * Created by www.muathu@gmail.com on 11/2/2017.
 */

class TypeWriterTextView : AppCompatTextView {
    private var mText: CharSequence? = null
    private var mIndex: Int = 0
    private var mDelay: Long = 150 // in ms

    private val mHandler = Handler()

    val characterAdder: Runnable = run {
        Runnable {
            mText?.let {
                text = it.subSequence(0, mIndex++)
                if (mIndex <= it.length) {
                    mHandler.postDelayed(characterAdder, mDelay)
                }
            }
        }
    }

    constructor(context: Context) : super(context) {}

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {}

    fun animateText(txt: CharSequence) {
        mText = txt
        mIndex = 0

        text = ""
        mHandler.removeCallbacks(characterAdder)
        mHandler.postDelayed(characterAdder, mDelay)
    }

    fun setCharacterDelay(m: Long) {
        mDelay = m
    }
}